src/ring_theory/integral_closure.lean

```diff
@@ -0,0 +1,279 @@
+/-
+Copyright (c) 2019 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Integral closure of a subring.
+-/
+
+import ring_theory.adjoin
+
+universes u v
+
+open polynomial submodule
+
+variables (R : Type u) {A : Type v}
+variables [comm_ring R] [comm_ring A]
+variables [decidable_eq R] [decidable_eq A]
+variables [algebra R A]
+
+def is_integral (x : A) : Prop :=
+∃ p : polynomial R, monic p ∧ aeval R A x p = 0
+
+variables {R}
+theorem is_integral_algebra_map {x : R} : is_integral R (algebra_map A x) :=
+⟨X - C x, monic_X_sub_C _,
+by rw [alg_hom.map_sub, aeval_def, aeval_def, eval₂_X, eval₂_C, sub_self]⟩
+
+theorem is_integral_of_subring {x : A} (T : set R) [is_subring T]
+  (hx : is_integral T (algebra.comap.to_comap T R A x)) : is_integral R (x : A) :=
+let ⟨p, hpm, hpx⟩ := hx in
+⟨⟨p.support, λ n, (p.to_fun n).1,
+  λ n, finsupp.mem_support_iff.trans (not_iff_not_of_iff
+    ⟨λ H, have _ := congr_arg subtype.val H, this,
+    λ H, subtype.eq H⟩)⟩,
+have _ := congr_arg subtype.val hpm, this, hpx⟩
+
+theorem is_integral_iff_is_integral_closure_finite {r : A} :
+  is_integral R r ↔ ∃ s : set R, s.finite ∧
+    is_integral (ring.closure s) (algebra.comap.to_comap (ring.closure s) R A r) :=
+begin
+  split; intro hr,
+  { rcases hr with ⟨p, hmp, hpr⟩,
+    exact ⟨_, set.finite_mem_finset _, p.restriction, subtype.eq hmp, hpr⟩ },
+  rcases hr with ⟨s, hs, hsr⟩,
+  exact is_integral_of_subring _ hsr
+end
+
+theorem fg_madjoin_singleton_of_integral (x : A) (hx : is_integral R x) :
```

Comment:
`to_submodule` is now a coercion. Also, I think it makes sense to leave the `m` out of `madjoin`.

Reply:
done